code_relocation: Add NOCOPY feature

carlocaione opened the pull request:
### Use case of interest:

Some platforms are shipping in parallel to the internal FLASH some other storage / external FLASH (usually a QSPI FLASH) that can be used to execute (XIP) code from.

The content of this external FLASH can usually be written at flash time using proper tools (see for example the case of the external FLASH on the nRF5340DK that can be written at flash time using nrfjprog).

The external FLASH is a nice addition that is extremely useful when a large application code doesn't entirely fit on the internal FLASH so that we could want to move part of it in the auxiliary FLASH to XIP the code from there.

### The problem:

Right now Zephyr doesn't have a formal and generic way to move at build time part of the code to a different memory region.

### The current status:

Zephyr is indeed shipping a `code_relocation` feature but that doesn't entirely match our needs.

When XIP is enabled, the `code_relocation` feature is used in Zephyr to move the selected code (that is to copy `.text` section, to initialize `.data` and zero `.bss`) from FLASH to SRAM at run time and execute code from SRAM.

The relocation is done by a generated snippet of code that is `memcpy()`-ing the right content to the destination region and also using some build-time generated portions of linker script to retrieve start and destination addresses and regions.

### This patch:

This patch is leveraging the `code_relocation` code and adding a `NOCOPY` feature. This feature is using the `code_relocation` feature to dynamically build the linker script snippets but entirely skipping the run-time code relocation so that the code can be XIP-ed from the destination region.

### Example:

Let's say that we have a big file called `huge_file.c` that we want to XIP from the external FLASH that is mapped in an EXTFLASH region.

In this case we should enable `CONFIG_XIP` and `CONFIG_CODE_DATA_RELOCATION` and instruct cmake as follows:
```
  zephyr_code_relocate(src/huge_file.c EXTFLASH_TEXT NOCOPY)
  zephyr_code_relocate(src/huge_file.c SRAM_DATA)
```
this means that:

- The `.text` section of the `huge_file.c` must reside in the EXTFLASH memory region and we do not need to copy the section there because we are going to XIP it (and we assume that the file is going to be placed in the external FLASH at flash time).
- The `.data` section of the `huge_file.c` must still reside in the SRAM memory region.

### TODOs:

It's is desirable to have the possibility to relocate libraries and pre-build files instead of source code files.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
code_relocation: Add NOCOPY feature

* Use case of interest:

Some platforms are shipping in parallel to the internal FLASH some other
storage / external FLASH (usually a QSPI FLASH) that can be used to
execute (XIP) code from.

The content of this external FLASH can usually be written at flash time
using proper tools (see for example the case of the external FLASH on
the nRF5340DK that can be written at flash time using nrfjprog).

The external FLASH is a nice addition that is extremely useful when a
large application code doesn't entirely fit on the internal FLASH so
that we could want to move part of it in the auxiliary FLASH to XIP the
code from there.

* The problem:

Right now Zephyr doesn't have a formal and generic way to move at build
time part of the code to a different memory region.

* The current status:

Zephyr is indeed shipping a code_relocation feature but that doesn't
entirely match our needs.

When XIP is enabled, the code_relocation feature is used in Zephyr to
move the selected code (that is to copy text section, to initialize data
and zero bss) from FLASH to SRAM at run time and execute code from SRAM.

The relocation is done by a generated snippet of code that is
memcpy()-ing the right content to the destination region also using some
build-time generated portions of linker script to retrieve start and
destination addresses and regions.

* This patch:

This patch is leveraging the code_relocation code and adding a NOCOPY
feature. This feature is using the code_relocation feature to
dynamically build the linker script snippets but entirely skipping the
run-time code relocation so that the code can be XIP-ed from the
destination region.

* Example:

Let's say that we have a big file called `huge_file.c` that we want to
XIP from the external FLASH that is mapped in an EXTFLASH region.

In this case we should enable `CONFIG_XIP` and
`CONFIG_CODE_DATA_RELOCATION` and instruct cmake as follows:

  zephyr_code_relocate(src/huge_file.c EXTFLASH_TEXT NOCOPY)
  zephyr_code_relocate(src/huge_file.c SRAM_DATA)

this means that:

- The .text section of the `huge_file.c` must reside in the EXTFLASH
  memory region and we do not need to copy the section there because we
  are going to XIP it (and we assume that the file is going to be placed
  in the external FLASH at flash time).
- The .data section of the `huge_file.c` must still reside in the SRAM
  memory region.

* TODOs:

It's desirable to have the possibility to relocate libraries and
pre-build files instead of source code files.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

SebastianBoe commented:
I suspect the architecture would be simpler if code_relocate used feature X that changed the VMA of a source file.

And then for the external flash use-case we use feature X directly.

Instead of having one big blob of infrastucture that now is poorly named and can do all kinds of things based on feature flags.

SebastianBoe commented:
Are there any existing samples in Zephyr that we could port to use this to prove that it is flexible enough for the real world ?

I assume someone must have been using external flash already at this point?

carlocaione commented:
> Are there any existing samples in Zephyr that we could port to use this to prove that it is flexible enough for the real world ?
> I assume someone must have been using external flash already at this point?

I can probably try to extend the sample to the nRF5340 using the external flash on the QSPI.

carlocaione committed:
code_relocation: Extend the nocopy sample to nRF5340dk

The nRF5340dk is shipping an external QSPI flash that can be used to do
XIP from. Extend the code_relocation nocopy sample to support this
platform so that part of the functions are executed from internal flash
and others from external flash.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented on doc/guides/code-relocation.rst:
```diff
@@ -91,12 +91,32 @@ This section shows additional configuration options that can be set in
 Sample
 ======
-A sample showcasing this feature is provided at
+A sample showcasting this feature is provided at
```
well, that's embarrassing :)

npitre commented:
A few comments:

```
MEMORY
{
     EXTFLASH (wx) : ORIGIN = 0x10000000, LENGTH = 0x800000
}
```

The "wx" attributes look wrong for flash. It should rather be "rx", no?
Same goes for existing memory entryes you didn't modify.

The following commit log text is confusing:

```
  When XIP is enabled, the code_relocation feature is used in Zephyr to
  move the selected code (that is to copy text section, to initialize data
  and zero bss) from FLASH to SRAM at run time and execute code from SRAM.
```

Isn't the purpose of XIP to leave .textsection in place and execute them 
from there? There may be legitimate reasons for wanting to execute (part
of) the code from SRAM, but that's not what XIP usually means. Clarification
would be nice.

Also a typo: `s/It's is desirable/It is desirable/`.

npitre reviewed (approved):
Other than minor nits this looks good.
